// ngx-stripe/esm2015/lib/services/stripe.service.js
import { __decorate, __param } from "tslib";
import { Injectable, Inject } from '@angular/core';
import { WindowRef } from './window-ref.service';
import { LazyStripeAPILoader, Status } from './api-loader.service';
import { STRIPE_PUBLISHABLE_KEY, STRIPE_OPTIONS } from '../interfaces/stripe';
import { StripeInstance } from './stripe-instance.class';
import { filter, map } from 'rxjs/operators';
let StripeService = class StripeService {
    constructor(key, options, loader, window) {
        this.key = key;
        this.options = options;
        this.loader = loader;
        this.window = window;
        if (key) {
            this.stripe = new StripeInstance(this.loader, this.window, key, options);
        }
    }
    getStripeReference() {
        return this.loader.asStream().pipe(filter((status) => status.loaded === true), map(() => this.window.getNativeWindow().Stripe));
    }
    getInstance() {
        return this.stripe.getInstance();
    }
    setKey(key, options) {
        return this.changeKey(key, options);
    }
    changeKey(key, options) {
        this.stripe = new StripeInstance(this.loader, this.window, key, options);
        return this.stripe;
    }
    elements(options) {
        return this.stripe.elements(options);
    }
    createToken(a, b) {
        return this.stripe.createToken(a, b);
    }
    createSource(a, b) {
        return this.stripe.createSource(a, b);
    }
    retrieveSource(source) {
        return this.stripe.retrieveSource(source);
    }
    paymentRequest(options) {
        return this.stripe.paymentRequest(options);
    }
    handleCardPayment(a, b, c) {
        return this.stripe.handleCardPayment(a, b, c);
    }
    handleCardAction(a) {
        return this.stripe.handleCardAction(a);
    }
    createPaymentMethod(a, b, c) {
        return this.stripe.createPaymentMethod(a, b, c);
    }
    confirmPaymentIntent(a, b) {
        return this.stripe.confirmPaymentIntent(a, b);
    }
    handleCardSetup(a, b, c) {
        return this.stripe.handleCardSetup(a, b, c);
    }
};
StripeService.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [STRIPE_PUBLISHABLE_KEY,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [STRIPE_OPTIONS,] }] },
    { type: LazyStripeAPILoader },
    { type: WindowRef }
];
StripeService = __decorate([
    Injectable(),
    __param(0, Inject(STRIPE_PUBLISHABLE_KEY)),
    __param(1, Inject(STRIPE_OPTIONS))
], StripeService);
export { StripeService };
