// @nomadreservations/ngx-stripe/esm2015/projects/ngx-stripe/lib/stripe-card.component.js
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { BehaviorSubject, combineLatest as observableCombineLatest } from 'rxjs';
import { filter, switchMap } from 'rxjs/operators';
import { StripeService } from './services/stripe.service';
let StripeCardComponent = class StripeCardComponent {
    constructor(stripeService) {
        this.stripeService = stripeService;
        this.change = new EventEmitter();
        this.complete = new EventEmitter();
        this.error = new EventEmitter();
        this.options$ = new BehaviorSubject({});
        this.elementsOptions$ = new BehaviorSubject({});
    }
    set options(optionsIn) {
        this.options$.next(optionsIn);
    }
    set elementsOptions(optionsIn) {
        this.elementsOptions$.next(optionsIn);
    }
    ngOnInit() {
        const elements$ = this.elementsOptions$.asObservable().pipe(switchMap(options => {
            if (Object.keys(options).length > 0) {
                return this.stripeService.elements(options);
            }
            return this.stripeService.elements();
        }));
        observableCombineLatest(elements$, this.options$.asObservable().pipe(filter(options => Boolean(options)))).subscribe(([elements, options]) => {
            if (this.card) {
                this.element = elements.create('card', options);
                this.element.mount(this.card.nativeElement);
                this.element.on('change', changedCard => {
                    this.change.emit({
                        card: changedCard,
                        element: this.element
                    });
                    if (changedCard.complete) {
                        this.complete.emit({
                            card: changedCard,
                            element: this.element
                        });
                    }
                    if (changedCard.error) {
                        this.error.emit(changedCard.error);
                    }
                });
            }
        });
    }
    getCard() {
        return this.element;
    }
};
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], StripeCardComponent.prototype, "change", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], StripeCardComponent.prototype, "complete", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", Object)
], StripeCardComponent.prototype, "error", void 0);
tslib_1.__decorate([
    ViewChild('card'),
    tslib_1.__metadata("design:type", ElementRef)
], StripeCardComponent.prototype, "card", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], StripeCardComponent.prototype, "options", null);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], StripeCardComponent.prototype, "elementsOptions", null);
StripeCardComponent = tslib_1.__decorate([
    Component({
        selector: 'ngx-stripe-card',
        template: `<div class="field" #card></div>`
    }),
    tslib_1.__metadata("design:paramtypes", [StripeService])
], StripeCardComponent);
export { StripeCardComponent };
