// input-mask-angular/esm2015/lib/input-mask-angular.directive.js
import { Directive, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
import { MaskedTextChangedListener } from 'ts-input-mask';
import { InputMaskOptions } from './input-mask-options';
import * as i0 from "@angular/core";
export class InputMaskAngularDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.maskFilled = new EventEmitter();
        this.extractedValue = new EventEmitter();
        this.formattedText = new EventEmitter();
        this.placeholder = new EventEmitter();
        this._options = new InputMaskOptions();
    }
    set value(value) {
        this._value = value;
    }
    set primaryFormat(value) {
        this._primaryFormat = value;
    }
    set options(value) {
        this._options = value;
    }
    ngOnInit() {
        this.setupListener(this.elementRef.nativeElement);
    }
    setupListener(input) {
        if (!!this._primaryFormat) {
            const emitChanges = (maskFilled, extractedValue, formattedText) => {
                this.maskFilled.emit(maskFilled);
                this.extractedValue.emit(extractedValue);
                this.formattedText.emit(formattedText);
            };
            const listener = MaskedTextChangedListener.installOn(this._primaryFormat, input, new class {
                onTextChanged(maskFilled, extractedValue, formattedText) {
                    emitChanges(maskFilled, extractedValue, formattedText);
                }
            }(), this._options.affineFormats, this._options.customNotations, this._options.affinityCalculationStrategy, this._options.autocomplete);
            this.renderer.setProperty(input, 'placeholder', String(listener.placeholder()));
            if (!!this._value) {
                listener.setText(this._value);
            }
            this.placeholder.emit(String(listener.placeholder()));
        }
        else {
            input.addEventListener('input', () => {
                this.extractedValue.emit(input.value);
            });
        }
    }
}
/** @nocollapse */ InputMaskAngularDirective.ɵfac = function InputMaskAngularDirective_Factory(t) { return new (t || InputMaskAngularDirective)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.Renderer2)); };
/** @nocollapse */ InputMaskAngularDirective.ɵdir = i0.ɵɵdefineDirective({ type: InputMaskAngularDirective, selectors: [["input", "mask", ""]], inputs: { value: "value", primaryFormat: ["mask", "primaryFormat"], options: "options" }, outputs: { maskFilled: "maskFilled", extractedValue: "extractedValue", formattedText: "formattedText", placeholder: "placeholder" } });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(InputMaskAngularDirective, [{
        type: Directive,
        args: [{
                selector: 'input[mask]'
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, { maskFilled: [{
            type: Output
        }], extractedValue: [{
            type: Output
        }], formattedText: [{
            type: Output
        }], placeholder: [{
            type: Output
        }], value: [{
            type: Input,
            args: ['value']
        }], primaryFormat: [{
            type: Input,
            args: ['mask']
        }], options: [{
            type: Input,
            args: ['options']
        }] }); })();
