// ts-input-mask/dist/util/string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
String.prototype.isDigit = function () {
    return this.search(/^[0-9]$/g) !== -1;
};
String.prototype.isLetter = function () {
    return this.search(/^[0-9]$/g) === -1;
};
String.prototype.isLetterOrDigit = function () {
    return this.isDigit() || this.isLetter();
};
String.prototype.contains = function (char) {
    return this.search(char) !== -1;
};
String.prototype.isEmpty = function () {
    return this.length === 0;
};
String.prototype.first = function () {
    return this.charAt(0);
};
String.prototype.toCharArray = function () {
    return this.split('');
};
String.prototype.prefixIntersection = function (another) {
    if (this.isEmpty() || another.isEmpty())
        return '';
    let endIndex = 0;
    while (endIndex < this.length && endIndex < another.length) {
        if (this[endIndex] === another[endIndex]) {
            ++endIndex;
        }
        else {
            return this.substring(0, endIndex);
        }
    }
    return this.substring(0, endIndex);
};
