// ts-input-mask/dist/helper/affinity-calculation-strategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const model_1 = require("../model");
require("../util/string");
var AffinityCalculationStrategy;
(function (AffinityCalculationStrategy) {
    AffinityCalculationStrategy[AffinityCalculationStrategy["WHOLE_STRING"] = 0] = "WHOLE_STRING";
    AffinityCalculationStrategy[AffinityCalculationStrategy["PREFIX"] = 1] = "PREFIX";
})(AffinityCalculationStrategy = exports.AffinityCalculationStrategy || (exports.AffinityCalculationStrategy = {}));
class AffinityCalculation {
    constructor(strategy) {
        this.strategy = strategy;
    }
    calculateAffinityOfMask(mask, text, autocomplete) {
        switch (this.strategy) {
            case AffinityCalculationStrategy.WHOLE_STRING: {
                return mask.apply(new model_1.CaretString(text.string, text.caretPosition), autocomplete).affinity;
            }
            case AffinityCalculationStrategy.PREFIX: {
                return mask.apply(new model_1.CaretString(text.string, text.caretPosition), autocomplete).formattedText.string.prefixIntersection(text.string).length;
            }
        }
    }
}
exports.AffinityCalculation = AffinityCalculation;
