// input-mask-angular/esm5/lib/input-mask-angular.pipe.js
import { Pipe } from '@angular/core';
import { CaretString, Mask } from 'ts-input-mask';
import * as i0 from "@angular/core";
var InputMaskAngularPipe = /** @class */ (function () {
    function InputMaskAngularPipe() {
    }
    InputMaskAngularPipe.prototype.transform = function (value, primaryFormat, customNotations) {
        if (customNotations === void 0) { customNotations = []; }
        if (!!primaryFormat) {
            var mask = Mask.getOrCreate(primaryFormat, customNotations);
            var stringLength = value.length;
            var result = mask.apply(new CaretString(value, stringLength), false);
            return String(result.formattedText.string);
        }
        return value;
    };
    /** @nocollapse */ InputMaskAngularPipe.ɵfac = function InputMaskAngularPipe_Factory(t) { return new (t || InputMaskAngularPipe)(); };
    /** @nocollapse */ InputMaskAngularPipe.ɵpipe = i0.ɵɵdefinePipe({ name: "mask", type: InputMaskAngularPipe, pure: true });
    return InputMaskAngularPipe;
}());
export { InputMaskAngularPipe };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(InputMaskAngularPipe, [{
        type: Pipe,
        args: [{
                name: 'mask'
            }]
    }], null, null); })();
