// input-mask-angular/esm5/lib/input-mask-angular.module.js
import { NgModule } from '@angular/core';
import { InputMaskAngularDirective } from './input-mask-angular.directive';
import { InputMaskAngularPipe } from './input-mask-angular.pipe';
import * as i0 from "@angular/core";
var InputMaskAngularModule = /** @class */ (function () {
    function InputMaskAngularModule() {
    }
    /** @nocollapse */ InputMaskAngularModule.ɵmod = i0.ɵɵdefineNgModule({ type: InputMaskAngularModule });
    /** @nocollapse */ InputMaskAngularModule.ɵinj = i0.ɵɵdefineInjector({ factory: function InputMaskAngularModule_Factory(t) { return new (t || InputMaskAngularModule)(); } });
    return InputMaskAngularModule;
}());
export { InputMaskAngularModule };
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(InputMaskAngularModule, { declarations: [InputMaskAngularDirective, InputMaskAngularPipe], exports: [InputMaskAngularDirective, InputMaskAngularPipe] }); })();
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(InputMaskAngularModule, [{
        type: NgModule,
        args: [{
                declarations: [InputMaskAngularDirective, InputMaskAngularPipe],
                exports: [InputMaskAngularDirective, InputMaskAngularPipe]
            }]
    }], null, null); })();
