// input-mask-angular/esm5/lib/input-mask-angular.directive.js
import { Directive, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
import { MaskedTextChangedListener } from 'ts-input-mask';
import { InputMaskOptions } from './input-mask-options';
import * as i0 from "@angular/core";
var InputMaskAngularDirective = /** @class */ (function () {
    function InputMaskAngularDirective(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.maskFilled = new EventEmitter();
        this.extractedValue = new EventEmitter();
        this.formattedText = new EventEmitter();
        this.placeholder = new EventEmitter();
        this._options = new InputMaskOptions();
    }
    Object.defineProperty(InputMaskAngularDirective.prototype, "value", {
        set: function (value) {
            this._value = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputMaskAngularDirective.prototype, "primaryFormat", {
        set: function (value) {
            this._primaryFormat = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputMaskAngularDirective.prototype, "options", {
        set: function (value) {
            this._options = value;
        },
        enumerable: true,
        configurable: true
    });
    InputMaskAngularDirective.prototype.ngOnInit = function () {
        this.setupListener(this.elementRef.nativeElement);
    };
    InputMaskAngularDirective.prototype.setupListener = function (input) {
        var _this = this;
        if (!!this._primaryFormat) {
            var emitChanges_1 = function (maskFilled, extractedValue, formattedText) {
                _this.maskFilled.emit(maskFilled);
                _this.extractedValue.emit(extractedValue);
                _this.formattedText.emit(formattedText);
            };
            var listener = MaskedTextChangedListener.installOn(this._primaryFormat, input, new /** @class */ (function () {
                function class_1() {
                }
                class_1.prototype.onTextChanged = function (maskFilled, extractedValue, formattedText) {
                    emitChanges_1(maskFilled, extractedValue, formattedText);
                };
                return class_1;
            }())(), this._options.affineFormats, this._options.customNotations, this._options.affinityCalculationStrategy, this._options.autocomplete);
            this.renderer.setProperty(input, 'placeholder', String(listener.placeholder()));
            if (!!this._value) {
                listener.setText(this._value);
            }
            this.placeholder.emit(String(listener.placeholder()));
        }
        else {
            input.addEventListener('input', function () {
                _this.extractedValue.emit(input.value);
            });
        }
    };
    /** @nocollapse */ InputMaskAngularDirective.ɵfac = function InputMaskAngularDirective_Factory(t) { return new (t || InputMaskAngularDirective)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.Renderer2)); };
    /** @nocollapse */ InputMaskAngularDirective.ɵdir = i0.ɵɵdefineDirective({ type: InputMaskAngularDirective, selectors: [["input", "mask", ""]], inputs: { value: "value", primaryFormat: ["mask", "primaryFormat"], options: "options" }, outputs: { maskFilled: "maskFilled", extractedValue: "extractedValue", formattedText: "formattedText", placeholder: "placeholder" } });
    return InputMaskAngularDirective;
}());
export { InputMaskAngularDirective };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(InputMaskAngularDirective, [{
        type: Directive,
        args: [{
                selector: 'input[mask]'
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }]; }, { maskFilled: [{
            type: Output
        }], extractedValue: [{
            type: Output
        }], formattedText: [{
            type: Output
        }], placeholder: [{
            type: Output
        }], value: [{
            type: Input,
            args: ['value']
        }], primaryFormat: [{
            type: Input,
            args: ['mask']
        }], options: [{
            type: Input,
            args: ['options']
        }] }); })();
